Move AKE API to deps section (#270)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

protocol which satisfies the forward secrecy and KCI properties discussed in
{{security-considerations}}.

> The AKE must define three messages `AKE_Init`, `AKE_Response` and `AKE_Finalize`

Reviewer: ```suggestion
The AKE must define three messages `AuthInit`, `AuthResponse` and `AuthFinish`
```

and provide the following functions for the client:

> - Start(): Initiate the AKE by producing message `AuthInit`.

Reviewer: Maybe make this `ClientStart`?

> - ClientFinalize(client_identity, client_private_key,

Reviewer: ```suggestion
- ClientFinish(client_identity, client_private_key,
```

> server_identity, server_public_key, `AKE_Init`): upon reception of the server's

Reviewer: ```suggestion
server_identity, server_public_key, `AKE_Init`): upon receipt of the server's
```
(and below for the server piece too)

response `AuthResponse`, complete the protocol for the client, produce
`AuthFinish`.

The AKE protocol must provide the following functions for the server:

> - Response(server_identity, server_private_key, client_identity,

Reviewer: Maybe make this `ServerStart`?

> client_public_key, `AKE_Init`): upon reception of a client's request `AKE_Init`,

Reviewer: ```suggestion
client_public_key, `AKE_Init`): upon receipt of a client's request `AKE_Init`,
```

engage in the AKE.

> - ServerFinish(`AKE_Finalize`): upon reception of a client's final AKE message

Reviewer: ```suggestion
- ServerFinish(`AKE_Finalize`): upon receipt of a client's final AKE message
```

`AuthFinish`, complete the protocol for the server.

> Both ClientFinalize and ServerFinish return an error if authentication failed.

Reviewer: ```suggestion
Both ClientFinish and ServerFinish return an error if authentication failed.
```

                        response
              <-------------------------

 (record, export_key) = FinalizeRequest(response,
                                        server_identity,
                                        client_identity)

                        record
              ------------------------->
~~~

> The client may utilize a state structure `RegistrationState` with the following

Reviewer: Can we revert this state change? It's separate from the AKE abstraction changes and (I think) sort of complicates presentation now.

Author: Do you prefer it in another PR or totally absent?

Reviewer: Right now, I lean towards it being totally absent, but we can always re-assess in a followup PR.

Author: Alright. I need to adapt the API to ask for the arguments that would otherwise be in the state

Author: Reverted 👍

                       session_key = ServerFinish(ke3)
~~~

Both client and server may use implicit internal state objects to keep necessary
material for the OPRF and AKE, `client_state` and `server_state`, respectively.

The client state may have the following named fields:

- password, the input password; and
- blind, the random blinding scalar returned by `Blind()`, of length Nok; and
- client_ake_state, the client's AKE state if necessary.

The server state may have the following fields:

- server_ake_state, the server's AKE state if necessary.

The rest of this section describes these authenticated key exchange messages
and their parameters in more detail. {{cred-retrieval}} discusses internal
functions used for retrieving client credentials, and {{ake-protocol}} discusses
how these functions are used to execute the authenticated key exchange protocol.

> ## External Client API {#opaque-client}

Reviewer: ```suggestion
## Client Authentication Functions {#opaque-client}
```

~~~
ClientInit(password)

State:
- state, a ClientState structure.

Input:
- password, an opaque byte string containing the client's password.

Output:
- ke1, a KE1 message structure.

Steps:
1. request, blind = CreateCredentialRequest(password)
2. state.blind = blind
3. ake_1 = Start(request)
4. Output KE1(request, ake_1)
~~~

~~~
ClientFinish(client_identity, server_identity, ke2)

State:
- state, a ClientState structure

Input:
- client_identity, the optional encoded client identity, which is set
  to client_public_key if not specified.
- server_identity, the optional encoded server identity, which is set
  to server_public_key if not specified.
- ke2, a KE2 message structure.

Output:
- ke3, a KE3 message structure.
- session_key, the session's shared secret.

> - export_key, , an additional client key.

Reviewer 2: extra comma here

Author: thanks! fixed :)

Steps:
1. (client_private_key, server_public_key, export_key) =
    RecoverCredentials(state.password, state.blind, ke2.CredentialResponse,
                       server_identity, client_identity)
2. (ke3, session_key) =
    ClientFinalize(client_identity, client_private_key, server_identity,
                    server_public_key, ke2)
3. Output (ke3, session_key)
~~~

> ## External Server API {#opaque-server}

Reviewer: ```suggestion
## Server Authentication Functions {#opaque-server}
```

~~~
ServerInit(server_identity, server_private_key, server_public_key,
           record, credential_identifier, oprf_seed, ke1)

Input:
- server_identity, the optional encoded server identity, which is set to
  server_public_key if nil.
- server_private_key, the server's private key.
- server_public_key, the server's public key.
- record, the client's RegistrationRecord structure.
- credential_identifier, an identifier that uniquely represents the credential.
- oprf_seed, the server-side seed of Nh bytes used to generate an oprf_key.
- ke1, a KE1 message structure.

Output:
- ke2, a KE2 structure.

Steps:
1. response = CreateCredentialResponse(ke1.request, server_public_key, record,
    credential_identifier, oprf_seed)
2. ake_2 = Response(server_identity, server_private_key,
    client_identity, record.client_public_key, ke1, response)
3. Output KE2(response, ake_2)
~~~

Since the OPRF is a two-message protocol, KE3 has no element of the OPRF. We can

> therefore call the AKE's `ServerFinish()` directly and return its output.

Reviewer: Don't we still need a function to call to get the `session_key` on the server side? (ServerFinish is the function for this, but I don't see why it's defined way below and not here?)

Author: Yeah, but I was hesitant in describing it here, since this part is generic without any details about the AKE. Specifying ServerFinish here will start talking about 3DH before the dedicated AKE section (if that makes sense)

Reviewer: Well, wouldn't _any_ AKE need this final function on the server? How does the server get the session key otherwise?

Author: Of course, but the internals of this function differ from one AKE to the other

Author: You're right, the function should still be described here. I described the necessary behaviour and API constraints of ServerFinish() in general. What do you think of that update?
